### ISSUE TITLE
Convert images to jpg/png to make them printable

### ISSUE THREAD

**Issue by legoktm** [freedomofpress/securedrop-client#2177]
## Description

Like we use LibreOffice to convert documents to PDFs so they can be printed, we should do the same for images.

## How will this impact [SecureDrop users](https://github.com/freedomofpress/securedrop-ux/wiki/Users)?

* Broaden the set of printable file types to include most images

## How would this affect the SecureDrop Workstation [threat model](https://github.com/freedomofpress/securedrop-workstation/#threat-model)?

* Likely introduce a new dependency on an image conversion tool, probably `imagemagick`. This tool runs inside a dispVM, so I don't think it would introduce any more attack surface than LibreOffice would.
  * @rocodes also mentioned that another option might be to use Dangerzone as the conversion mechanism.

## User Stories

* As a journalist, I want to print webp/gif/raw/etc. files, so that I can share them offline.
* As a journalist, I want to be able to print images and not need to care about how the source sent them to me.

**Comment by rocodes** [comment]
Good things about including imagemagick in the large template, not just for printing:
- print more filetypes
- view more filetypes
- other potentially-useful built in functionality (image blur?)

I think ideally we'd be able to stop supporting multiple separate export paths, so we wouldn't just install imagemagick for printing purposes, but along a single consolidated print/export path (that might/should involve dangerzone)  :)

**Comment by legoktm** [comment]
Yeah, I feel that this is kind of a problem set that we might better delegate to Dangerzone instead of having two similar project/implementations. I don't know how far/close DZ is ready to being plugged in and if we're ready for that.

----

I was poking at imagemagick's identify command that can be used to well, identify a file:
```
user@dev ~/g/f/s/e/t/f/samples_supported (main)> identify -verbose Sample_jpeg.jpg | grep -i mime
  Mime type: image/jpeg
```

But then if you try it on a docx:
```
user@dev ~/g/f/s/e/t/f/samples_supported (main)> identify -verbose Sample_docx.docx | grep -i mime
convert /tmp/magick-OvuQ7lXsiWvGrVV0SnLLYw6UMoAYvo50 as a Writer document -> /tmp/magick-OvuQ7lXsiWvGrVV0SnLLYw6UMoAYvo50.pdf using filter : writer_pdf_Export
  Mime type: application/pdf
  Mime type: application/pdf
```

Oops, it is "smart" and converts the docx to pdf and then says oh, it's a pdf. But how is it converting docx to pdf?

```
$ magick -list delegate

Path: /etc/ImageMagick-7/delegates.xml

Delegate                Command
-------------------------------------------------------------------------------
        doc =>          "libreoffice' --convert-to pdf -outdir `dirname '%i'` '%i' 2> '%u'; /usr/bin/mv '%i.pdf' '%o"
       docx =>          "libreoffice' --convert-to pdf -outdir `dirname '%i'` '%i' 2> '%u'; /usr/bin/mv '%i.pdf' '%o"
```
(trimmed because there's a bunch more formats)

It's just using libreoffice under the hood 🙃 

I couldn't find any way to turn off the conversion with `identify`, maybe if we just overwrite the delegates.xml file to strip out the libreoffice ones? Or...we just use imagemagick's wrapper over libreoffice and then only need to worry about imagemagick instead of also libreoffice?

**Comment by almet** [comment]
Heya,

So if I understand correctly, the goal would be to use dangerzone to convert images to a printable format.

This is currently doable with the command line utility (it only produces PDFs, no jpg/png as mentioned in the issue):

```bash
dangerzone-cli --output-filename output.pdf input.format
```

Some of the formats listed in the issue aren't supported (`webp`, for instance, is not). Here is the list of currently supported extensions (from the docs):

> - PDF (.pdf)
> - Microsoft Word (.docx, .doc)
> - Microsoft Excel (.xlsx, .xls)
> - Microsoft PowerPoint (.pptx, .ppt)
> - ODF Text (.odt)
> - ODF Spreadsheet (.ods)
> - ODF Presentation (.odp)
> - ODF Graphics (.odg)
> - Hancom HWP (Hangul Word Processor) (.hwp, .hwpx)
> 	-     Not supported on Qubes OS
> - EPUB (.epub)
> - Jpeg (.jpg, .jpeg)
> - GIF (.gif)
> - PNG (.png)
> - SVG (.svg)
> - other image formats (.bmp, .pnm, .pbm, .ppm)

That might be a good opportunity to look adding support for more formats.

---

> if the conversion were to happen in the viewer itself instead of spinning up an intermediate VM

I'm not sure to understand what you mean by this? (what do you call "the viewer" ?). Do you need the security guarantees provided by dangerzone?

**Comment by almet** [comment]
Thanks for the explanation, it makes more sense indeed :+1:

I have one question left: you're talking about Qubes here, and my understanding is that it's not the only "environment" on which securedrop-client can run. On these other environments, would you also need to use dangerzone (maybe in a more "traditional" way)?

Now, on Qubes, the part that's currently running in the DZ container (that you named "the server") can probably be resumed to the [`doc_to_pixels.py`](https://github.com/freedomofpress/dangerzone/blob/main/dangerzone/conversion/doc_to_pixels.py) file. 

I'm not sure if this use case should be handled by the dangerzone project directly, as it seem very specific to how Qubes work, but we should definitely make it possible by offering APIs that makes this as simple as possible.

**Comment by eloquence** [comment]
What's the expected behavior in SecureDrop Client 0.15.0? In my testing, I can print JPG, PNG, and TIFF.

Other image types (SVG, WEBP) raise an `ERROR_MIMETYPE_UNSUPPORTED`.

Other non-printable types raise "This file cannot be printed".

**Comment by legoktm** [comment]
https://github.com/freedomofpress/securedrop-client/blob/main/export/securedrop_export/print/service.py has a list of file types that CUPS can print directly (including JPG, PNG and TIFF). 

We also have dynamic discovery of all the file types LibreOffice supports: https://github.com/freedomofpress/securedrop-client/blob/e039e9f1db63444e7060c7eb4bc69be302ababcc/export/securedrop_export/print/service.py#L129 since there are like 100+ IIRC, but they're mostly document formats.

IIRC this task is about the remaining subset of image formats that CUPS doesn't support natively.

**Comment by eloquence** [comment]
For the rewrite, I would nominate this as a post-1.0.0 enhancement, but adding to next backlog review for final decision on that.
